FormSpec: 9-slice images, animated_images, and fgimg_middle

v-rob opened the pull request:
This PR closes #9649 by adding an extra optional parameter to `image` and `animated_image` defining the 9-slice rect for the (animated) image to use.  It does _not_ make `image9` and `animated_image9`; `background9` is just an alias to `background` to allow old clients to not display errors before formspec versioning existed.  In addition, it also adds the `fgimg_middle` style property for image buttons

![gif](https://user-images.githubusercontent.com/31123645/90200841-08be3880-dd8e-11ea-8278-0ffbcb977651.gif)

## To do

This PR is Ready for Review.

## How to test

The above screenshot:

```lua
[[
formspec_version[4]
size[9,9]
style_type[button;bgimg=air.png;bgimg_middle=1]
label[1,0.8;bgimg_middle]
button[1,1;3,3;;]
style_type[image_button;fgimg=air.png;fgimg_middle=1]
label[5,0.8;fgimg_middle]
image_button[5,1;3,3;;;]
image[1,5;3,3;air.png;2,1]
label[1,4.8;9-slice image]
animated_image[5,5;3,3;;]] .. minetest.formspec_escape("[combine:16x32:0,0=air.png:0,16=ignore.png") .. [[;2;1000;1;1]
label[5,4.8;9-slice animated_image]
]]
```

Also make sure that all rect types work.

v-rob committed:
FormSpec: 9-slice images and animated_images

v-rob committed:
Add fgimg_middle; clean up code

v-rob commented:
Added support for nine-slice image buttons with the `fgimg_middle` style property.

kilbith commented:
Can `fgimg_middle` be extended to `image_button`?

v-rob commented:
> Can `fgimg_middle` be extended to `image_button`?

It already is for `image_button[]`; normal buttons don't have `fgimg` support.  Does it not work, or was this a misunderstanding?

kilbith commented:
https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L2840

I don't see any mention of ``fgimg_middle` there.

v-rob commented:
Maybe because that links to `minetest/master`'s `lua_api.txt` :)  But you are correct, I do not have it documented.  Will do and rebase.

kilbith commented:
> But you are correct, I do not have it documented.

Nope, edited my previous comment.

kilbith commented:
Well using `fgimg_middle` does not seem to make any difference. Looks like unimplemented:

Without:
![Capture d’écran de 2020-12-31 22-59-39](https://user-images.githubusercontent.com/7883281/103426789-37fa0880-4bbc-11eb-809d-7fe170f2d55b.png)

With:
![Capture d’écran de 2020-12-31 22-59-25](https://user-images.githubusercontent.com/7883281/103426794-3c262600-4bbc-11eb-8d21-38c944ff6090.png)

v-rob commented:
Rebased and documented.  `fgimg_middle` seems to work fine for me:

Without:
![image](https://user-images.githubusercontent.com/31123645/103429672-59d19c00-4b78-11eb-9ffb-1626cdc88525.png)

With:
![image](https://user-images.githubusercontent.com/31123645/103429545-51796100-4b78-11eb-87a6-192245d6170a.png)

kilbith commented:
So I misunderstood, I thought you said it was already implemented in `upstream/master`...

v-rob commented:
> So I misunderstood, I thought you said it was already implemented in `upstream/master`...

Ah, I see.  I was just saying that the Lua API link you provided was to master's, not mine.  Good to have that cleared up.

rubenwardy commented on src/gui/guiAnimatedImage.h:
```diff
@@ -1,28 +1,41 @@
 	virtual void draw() override;
 
-	void setFrameIndex(s32 frame);
+	void setImage(video::ITexture *texture) { m_texture = texture; };
```
`setTexture`.  Same for getImage

rubenwardy commented on src/gui/guiButtonImage.cpp:
```diff
@@ -32,15 +32,15 @@ using namespace gui;
-			core::rect<s32>(0,0,rectangle.getWidth(),rectangle.getHeight()), this);
-	m_image->setScaleImage(isScalingImage());
+	GUIButton::setScaleImage(true);
+	m_image = new GUIAnimatedImage(environment, this, id, rectangle);
```
irr_ptr?

v-rob replied:
Not possible without changing a lot more code like `m_clickthrough_elements`, better leave it to dedicated PRs.

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -766,108 +766,75 @@ void GUIFormSpecMenu::parseScrollBarOptions(parserData* data, const std::string
 
-		MY_CHECKPOS("image", 0);
-		MY_CHECKGEOM("image", 1);
+	if (parts.size() < 3 || parts.size() > 4 ||
```
This is not correct.

Fix:

```c++
if ((parts.size() < 3 || parts.size() > 4) && (m_formspec_version <= FORMSPEC_API_VERSION || parts.size() < 3)) {
```

Also, this should really be a macro or something. I like how my FormspecElement does it

v-rob replied:
Yes, the FormspecElement was very nice.  I made something with `std::initializer_list` in my partial refactor like `checkSize(m_formspec_version, {3, 4})`, but that might count as too C++y.

rubenwardy replied:
Actually, I think the logic in this is clearer despite being longer:


```c++
if (!((parts.size() >= 3 && parts.size() <= 4) || (m_formspec_version > FORMSPEC_API_VERSION && parts.size() > 4))) {
```

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -766,108 +766,75 @@ void GUIFormSpecMenu::parseScrollBarOptions(parserData* data, const std::string
+		parseRect(parts[3], &middle);
 
-		v2s32 pos = getElementBasePos(&v_pos);
+	GUIAnimatedImage *e = new GUIAnimatedImage(Environment, data->current_parent,
```
Wouldn't it be better not to add the image if the texture is null?

v-rob replied:
Honestly, this could apply to a lot of elements, e.g. empty text on a `label` or `vertlabel`.  However, I'd personally leave these elements in; for instance, `model` can have a `bgcolor` set, and if the element wasn't created if the model provided didn't exist, the bgcolor wouldn't appear.  `image` and `animated_image` currently don't have such properties, but they could quite easily in the future.

rubenwardy commented on src/gui/guiAnimatedImage.cpp:
```diff
@@ -34,40 +20,43 @@ void GUIAnimatedImage::draw()
 	if (m_texture != nullptr) {
```
early exit would be better

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -766,108 +766,75 @@ void GUIFormSpecMenu::parseScrollBarOptions(parserData* data, const std::string
-		v2s32 pos;
-		v2s32 geom;
+	std::vector<std::string> v_pos = split(parts[0],',');
+	std::vector<std::string> v_geom = split(parts[1],',');
```
Doesn't this break compat for images without a size?

v-rob replied:
It's undocumented, and I've never seen it in any mods, so I removed it.

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -766,108 +766,75 @@ void GUIFormSpecMenu::parseScrollBarOptions(parserData* data, const std::string
-			!(parts.size() > 7 && m_formspec_version > FORMSPEC_API_VERSION)) {
-		errorstream << "Invalid animated_image element(" << parts.size()
-			<< "): '" << element << "'" << std::endl;
+	if (parts.size() < 6 || parts.size() > 8 ||
```
Same issue here

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1042,6 +1018,35 @@ void GUIFormSpecMenu::parseButton(parserData* data, const std::string &element,
 	errorstream<< "Invalid button element(" << parts.size() << "): '" << element << "'"  << std::endl;
 }
 
+bool GUIFormSpecMenu::parseRect(const std::string &value, core::rect<s32> *parsed_rect)
```
`parseMiddleRect` would be better imo

rubenwardy reviewed (commented):
Haven't tested yet

Also, please add an example to games/devtest/mods/testformspec/formspec.lua

v-rob committed:
Address issues, add tests

kilbith commented:
Tested. `fgimg_middle=*` makes my tabs squarred whereas these are rounded normally.

![Capture d’écran de 2021-01-27 20-17-58](https://user-images.githubusercontent.com/7883281/106042222-0aca4880-60dd-11eb-915b-7ec61af5ec94.png)

I am doing something wrong? https://github.com/minetest-mods/i3/blob/main/init.lua#L1972

v-rob commented:
Could you try `bgimg` and `bgimg_middle` (which is probably the correct choice for those buttons anyway; fgimg is mostly for images that go above another image instead of being the background themselves) and see if that works?  If that doesn't work, then it's likely your code.

kilbith commented:
Converting to `bgimg_*` makes it worse. The text is squashed and the tabs are still squarred.

![Capture d’écran de 2021-01-27 21-42-02](https://user-images.githubusercontent.com/7883281/106051470-c8a70400-60e8-11eb-80a2-c9c53542edce.png)

kilbith commented:
BTW, using `bgimg_*` on my tabs leaves a gray background around the corners.

![Capture d’écran de 2021-01-28 21-09-42](https://user-images.githubusercontent.com/7883281/106193110-5d733580-61ad-11eb-9492-7241a19227be.png)

So your tips are completely wrong, @v-rob. You should be modding more with the GUI tools and recognize that your code looks broken.

v-rob commented:
I don't know what you are doing, but it works almost exactly as expected when I use `bgimg`.  Firstly, the text looks cut off because `bgimg_middle` applies an offset to padding (it's a documented feature), so to remove that offset, use `padding` with the same offset, but negative.  For me, the nine-slice rect works fine:

```lua
fs(fmt([[style_type[image_button;bgimg=%s;bgimg_hovered=%s;bgimg_middle=13;padding=-13;noclip=true;
			font_size=16;textcolor=%s;content_offset=0;sound=i3_tab] ]],
```

![image](https://user-images.githubusercontent.com/31123645/106324238-abc81900-622d-11eb-850d-050f1abf6b09.png)

Even without `padding`, the nine-slice rect still works fine, although the text is cut off.  Of course, these buttons won't work well with nine-slice images because that ruins the gradient as seen in the screenshot, so you'll either have to use something other than a gradient or forgo the nine-slice images because they're incompatible.

However, the light gray behind the button is a bug with image buttons (which I will look into) and unrelated to this PR.  Using normal buttons (which support `bgimg` as well) should work fine for now, especially seeing as those buttons don't require `fgimg`s on them.

Therefore, there is no issue with this PR found so far, and my advice is not incorrect.

Df458 commented:
Ok, finally had a moment to do some testing. First thoughts--I highly appreciate the 9-slice images, they work well for me.

Second thought:
@kilbith is both right and wrong here. His comments above show a pretty clear misunderstanding of how 9-slice images work in formspecs, and the main reason i3's tabs don't work in 9-slice is that they are not correctly formatted as 9-slice textures. Remember--the entire point of 9-slice is *to scale the scale as little of the texture as possible*. This means the texture is not scaled down, which is why your tabs look wrong. Like @v-rob, I'm unable to repro the gray background at all... I'd need a more complete code example to see what's happening there.

**With all that said**, i3 *does* showcase a regression in 9-slice behavior caused by this PR.

Using the following code snippet at the end of `get_items_fs`:
```lua
		fs(fmt([[style_type[button;bgimg=%s;bgimg_hovered=%s;bgimg_middle=16,0,-16,-16;padding=-16,0,16,16;noclip=true;
			font_size=16;textcolor=%s;content_offset=0;sound=i3_tab] ]],
		selected and PNG.tab_small_hover or PNG.tab_small,
		PNG.tab_small_hover, selected and "#fff" or "#ddd"))

		fs("style_type[button:hovered;textcolor=#fff]")
		fs("button", (data.inv_width - 0.65) + (i * (tab_len + 0.1)),
			full_height, tab_len, tab_hgh, fmt("itab_%u", i), title)
```

...we can observe two different results:
![Screenshot-2021-08-21-18:51:05](https://user-images.githubusercontent.com/2219707/130336721-3c1b4bc9-be91-4236-89f2-bd153eca9b32.png)

That's all the digging I'm in the mood for today. Hopefully we can resolve any remaining issues quickly and get this highly useful PR merged!

AFCMS commented:
The 9 sliced image looks perfect! I was waiting for this feature for a long time and hope it will be merged soon!
![image](https://user-images.githubusercontent.com/61794590/131219557-c1e86bbb-bb43-4c01-96ff-0535ed601e44.png)

rubenwardy commented:
You should be able to use background9 to create that GUI in stable Minetest versions. With formspec version 3, there's not that much difference between images and background images

v-rob commented:
Actually, backgrounds are still drawn behind everything else while images can be layered anywhere.

rubenwardy commented:
The code looks good to me

@Df458: please may you provide a minimal reproducable example, the context isn't very clear to me

Df458 commented:
Sure, I'll see about getting a repro tomorrow.

Df458 commented:
One piping-hot repro, coming right up! I've made a very small formspec with a recolored texture to highlight the difference:
[9slice.zip](https://github.com/minetest/minetest/files/7456376/9slice.zip)

Here's a before-after comparison screenshot to compare as well:
![Screenshot-2021-11-01-17:57:05](https://user-images.githubusercontent.com/2219707/139747864-2d113d86-dfc8-482c-bad3-c37cbb407789.png)

v-rob commented:
That does appear problematic.  I'll have to look into it.

v-rob committed:
Fix stupid error; bump formspec version

v-rob commented:
Fixed the error.  It was a simple case of using a width where an height was needed :-|  I added that case to `devtest`.

Formspec version has been bumped to 5 since 4 was the last MT version.

Df458 commented:
Tested, looks like the regressions have been fixed!

Df458 commented on games/devtest/mods/testformspec/formspec.lua:
```diff
@@ -375,12 +377,16 @@ local pages = {
 	-- Animation
 		[[
-			formspec_version[3]
+			formspec_version[5]
```
```suggestion
			formspec_version[6]
```

Df458 commented:
Initial impression is good, but I haven't dug too deeply in the code or done any thorough testing yet.

v-rob commented:
Well, found a bug in GitHub: no reply field for the first review (although the rest have review fields):
![image](https://user-images.githubusercontent.com/31123645/163292947-7030b29c-cda6-46ef-8d01-ce502a2f9708.png)

Well, I'll reply here instead.  I think I agree; especially considering the kerfuffles that have arisen about backwards compatibility being broken without warning for undocumented things, it's a much better idea to leave it in.  I absolutely hate how mods (and even devtest!) are using undocumented behaviour, but it's a better idea to not break it for this sort of thing.  So, might as well just document it and leave it as another artifact of how garbage formspecs are.

Df458 commented:
Looked through the rest of the code, and checked the following content for visual defects:
- i3
- cartographer
- devtest
- labyrinth
- mineclone 2

Haven't seen any issues aside from the ones I pointed out previously, so barring any new revelations I will approve once those are resolved.

sfan5 commented:
@v-rob do you have time to work on this soon? (seeing as it's scheduled for 5.6)
alternatively perhaps @Df458 could

v-rob commented:
See #12453; sorry about that.
